sklearn/utils/isfinite.pyx

```diff
@@ -0,0 +1,122 @@
+# Author: jakirkham
+
+cimport libc
+cimport libc.math
+from libc.math cimport isfinite as _isfinite, isinf as _isinf
+
+cimport cython
+from cython.parallel import prange
+cimport numpy
+import numpy
+numpy.import_array()
+
+
+cdef fused fprecision:
+    float
+    double
+
+cdef fused const_fprecision:
+    const float
+    const double
+
+
+cdef enum bint_enum:
+    false = False
+    true = True
+
+
+cdef struct bint_false_type:
+    char empty
+
+cdef struct bint_true_type:
+    char empty
+
+cdef fused bint_type:
+    bint_false_type
+    bint_true_type
+
+
+@cython.cdivision(True)
+def cy_isfinite(numpy.ndarray a, bint allow_nan=False):
+    cdef numpy.ndarray a_flat
+
+    cdef char* a_data
+    cdef numpy.NPY_TYPES a_type
+    cdef Py_ssize_t a_step
+    cdef Py_ssize_t a_size
+
+    cdef bint disallow_nan
+
+    cdef bint result
+    cdef bint_enum err
+
+
+    a_flat = numpy.PyArray_Reshape(a, -1)
```

Comment:
If the array is contiguous, I do not think we need to reshape at all.

Reply:
Actually in general do we need the array to be C/F contiguous? We don't really care about accessing the data "in order" per se, so it should be fine without a contiguous array right? (I may be misunderstanding the contiguity of ndarrays)

Reply:
My overall concern is how reshape can copy and copying `X` is a huge memory cost to pay. Using `C/F` contiguous is one way to avoid the possible copy from reshape. If the array is contiguous, the data can be traversed without reshaping.

Also, when the array is contiguous, it can take advantage of data locality and benefit from having the data on the same CPU cache line.

Reply:
Gotcha -- I was initially talking about removing reshape and simultaneously not stipulating contiguity, but the cache argument makes sense.